File: Leetcode/Tree/230.kth-smallest-element-in-a-bst.java
/*
Given a binary search tree, write a function kthSmallest to find the kth smallest element in it.
 
Example 1:

Input: root = [3,1,4,null,2], k = 1
   3
  / \
 1   4
  \
   2
Output: 1

Example 2:

Input: root = [5,3,6,2,4,null,null,1], k = 3
       5
      / \
     3   6
    / \
   2   4
  /
 1
Output: 3

Follow up:
What if the BST is modified (insert/delete operations) often and you need to find the kth smallest frequently? How would you optimize the kthSmallest routine?
 
Constraints:

    The number of elements of the BST is between 1 to 10^4.
    You may assume k is always valid, 1 ≤ k ≤ BST's total elements.

********************************************************************************************Solution**************************************************************************************************/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution 
{
    int kthSmallest = -1;
    public int kthSmallest(TreeNode root, int k) 
    {
        inorder(root, k);
        
        return kthSmallest;    
    }
    
    int i = 0;
    void inorder(TreeNode root, int k)
    {
        if(root == null)
            return;
        
        inorder(root.left, k);
        
        if(kthSmallest != -1)
            return;
        
        i++;
        if(i == k)
        {
            kthSmallest = root.val;
            return;
        }
                            
        inorder(root.right, k);        
    }
}

File: DSA-JavaScript/Binary Search Tree/230.kth-smallest-element-in-a-bst.js
/*

Given the root of a binary search tree, and an integer k, return the kth smallest value (1-indexed) of all the 
values of the nodes in the tree.


Example 1:

Input: root = [3,1,4,null,2], k = 1
Output: 1

Example 2:

Input: root = [5,3,6,2,4,null,null,1], k = 3
Output: 3
 

Constraints:

The number of nodes in the tree is n.
1 <= k <= n <= 10^4
0 <= Node.val <= 10^4

************************************************************Solution**************************************************/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
const kthSmallest = (root, k) => {
  const nodes = [];

  function util(node = root) {
    if (node === null) {
      return;
    }

    util(node.left);
    nodes.push(node.val);
    util(node.right);
  }

  util();
  return nodes[k - 1];
};
